plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -474,33 +476,26 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          //    - or the net_plugin received a block.
          //    - or we have reached the read_window_deadline
          void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, fc::time_point deadline) {
-            assert(num_tasks > 0 && num_waiting == 0);
+            std::lock_guard<std::mutex> g( mtx ); // not strictly necessary with current usage from single thread
+            assert(num_tasks > 0 && num_waiting == 0 && received_block != nullptr);
             assert(received_block && *received_block == false);
             max_waiting = num_tasks;
             received_block_ptr = received_block;
             read_window_deadline = deadline;
          }
 
       private:
-         void notify_waiting(bool all) {
-            if (num_waiting) {
-               all ? cond.notify_all() : cond.notify_one();
-            }
-         }
-            
          bool should_exit() {
-            return *received_block_ptr || exiting_read_window || 
-               fc::time_point::now() >= read_window_deadline;
+            return *received_block_ptr || exiting_read_window || fc::time_point::now() >= read_window_deadline;
```

Comment:
I think the condition in #862 is better and clearly spelled out (`num_waiting == max_waiting || fc::time_point::now() >= read_window_deadline || *received_block_ptr`). And the order reflects the likelihood of conditions too.

Reply:
I don't want to test `num_waiting == max_waiting` here on purpose.